### ISSUE TITLE
Interpreting training results and showing loss graph, YOLOv5s6

### ISSUE THREAD

**Issue by Carolinejone** [ultralytics/yolov5#8185]
Hello, I'm new to AI. I would like to ask for some guidance on my research. I'm doing my master thesis with YOLOv5. I'm trying to detect the anomaly of stay cables of cable-type bridges. 
 I used two datasets; Dataset- 1 has 2049 source images and 4159 images in total after combining with some augmented images to the training dataset, and Dataset- 2 has 1823 source images and 3673 images in total after combining with some augmented images to the training dataset. Train, validation and test dataset split is as follow,
Dataset -1, 3.2k: 614: 380
Dataset -2, 2.9k: 552: 299 
I followed custom YOLOv5 training from Roboflow and also used the Roboflow annotator with rectangular bounding boxes. I used batch size 64 and epoch 100 with pre-trained weight YOLOv5s6 and trained on COLAB pro plus for both datasets and the rest is the same as the custom YOLOv5 training tutorial.
This is the result for Dataset-1
![results](https://user-images.githubusercontent.com/88965698/173272561-dc0951d6-8493-4118-9845-eefcf9f71e92.png)
This is the result for Dataset-2
![results](https://user-images.githubusercontent.com/88965698/173273466-32b8e156-96dc-4f7d-a463-ef40871936b0.png)

My question is,

1.  Is the 100 epochs enough to train the model for both datasets 1 and 2?.
2. How to interpret the result tables including confusion matrics? (I already learned about Precision, Recall, mAP and F1 graph)
3. Which evaluation matric should I use, mAP or F1 or both? (Cause I've read that for class imbalance problem F1 score is the appropriate metric even though I believe that my model does not suffer from class imbalance effect.)
4. Is my model good enough or is it overfitting or is it needed to improve? 
5. Is the validation loss the only way to check overfitting?
6. Please provide some useful articles or papers for reference.
Thank you so much for your help.
Edit: adding more questions:
7. What is the difference between the label graph from result tables and the class balance from the health check of Roboflow.
8. How to interpret the results of label graph?
Dataset-1 label graph
![labels](https://user-images.githubusercontent.com/88965698/173275397-3293777d-03cc-42e0-975a-517ae5d1f226.jpg)
Dataset-2 label graph
![labels](https://user-images.githubusercontent.com/88965698/173275420-549324c8-7d03-4f7b-a323-d3d5bb9b5651.jpg)
Dataset-1 Class Balance
![Capture](https://user-images.githubusercontent.com/88965698/173275618-cc44926a-f736-4d26-bed7-c56c872223d6.JPG)
Dataset-2 Class Balance
![Capture](https://user-images.githubusercontent.com/88965698/173275650-460eef47-70a3-40a1-9f6b-313f20d78539.JPG)

**Comment by glenn-jocher** [comment]
@Carolinejone 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. I would combine both your datasets into a single dataset for best results. Also you must train longer until you see overfitting, otherwise you have not trained long enough.

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by Carolinejone** [comment]
@glenn-jocher Thank you so much for your explanation. I intentionally separate the dataset into two as I would like to compare the impact on the model's accuracy by the two different test datasets having different drone distances from the target object. For that reason, I couldn't combine the two datasets to make it larger. 
I've tried to train the model for 300 epochs and the results show like this for dataset-1, the model overfits earlier. (I'm still running the dataset-2 but I'm sure the result will be the same).

!python train.py --img 640 --batch 64 --epochs 300 --data {dataset.location}/data.yaml --weights yolov5s6.pt --cache

Could you please explain how to prevent overfitting and if possible explain with some coding examples?

![R_curve](https://user-images.githubusercontent.com/88965698/173377167-3bed7288-cb7a-47a7-b181-63c47c9b0150.png)
![P_curve](https://user-images.githubusercontent.com/88965698/173377185-25eedf0c-b54f-491d-bc7c-38f6473252f2.png)
![PR_curve](https://user-images.githubusercontent.com/88965698/173377195-4bfb8510-11b1-4b33-b412-fb536aef994e.png)
![F1_curve](https://user-images.githubusercontent.com/88965698/173377207-ee2e4ef3-b196-4553-a1e0-e3ca555eba7e.png)

![results](https://user-images.githubusercontent.com/88965698/173377134-586c9e50-20e8-4bf8-b2d8-e83e266b8de8.png)
![confusion_matrix](https://user-images.githubusercontent.com/88965698/173377156-98198ff8-9bdb-4a4f-8991-7e1eafd114e4.png)

![labels](https://user-images.githubusercontent.com/88965698/173377107-3e6bfe29-f6f6-472f-9388-7bfe49b595c8.jpg)

**Comment by Carolinejone** [comment]
This is the updated result for dataset-2 after running 300 epochs using the same codes as above.
![confusion_matrix](https://user-images.githubusercontent.com/88965698/173382387-c92a2d3c-3d50-4c71-a8c3-24d10db480f6.png)
![F1_curve](https://user-images.githubusercontent.com/88965698/173382400-4f1c64ea-865c-4182-80f5-c2599c1f844c.png)
![labels](https://user-images.githubusercontent.com/88965698/173382447-684a0973-c026-4501-b222-24c88e94616b.jpg)
![P_curve](https://user-images.githubusercontent.com/88965698/173382470-7581db2e-2d67-4912-82dd-0d6496b4eb8c.png)
![PR_curve](https://user-images.githubusercontent.com/88965698/173382477-9dedf4b6-9f31-4651-90b4-f16a45fce800.png)
![R_curve](https://user-images.githubusercontent.com/88965698/173382491-af4f1349-3ed2-4f75-9c2f-6b213d680aba.png)
![results](https://user-images.githubusercontent.com/88965698/173382506-d7641f3e-cfec-4dca-a01c-504364970d6e.png)

**Comment by glenn-jocher** [comment]
@Carolinejone follow recommendations in our complete guide in https://github.com/ultralytics/yolov5/issues/8185#issuecomment-1153788390. 

To reduce overfitting you can try using higher augmentation yamls in data/hyps or customize with more Albumentations.

YOLOv5 🚀 applies online imagespace and colorspace augmentations in the trainloader (but not the val_loader) to present a new and unique augmented [Mosaic](https://github.com/ultralytics/yolov5/blob/90b7895d652c3bd3d361b2d6e9aee900fd67f5f7/utils/datasets.py#L678-L732) (original image + 3 random images) each time an image is loaded for training. **Images are never presented twice in the same way**.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120995721-f3cfed00-c785-11eb-8ee2-b6ef2fa205e8.jpg)

### Augmentation Hyperparameters

The hyperparameters used to define these augmentations are in your hyperparameter file (default `data/hyp.scratch.yaml`) defined when training:
```
python train.py --hyp hyp.scratch-low.yaml
```
https://github.com/ultralytics/yolov5/blob/b94b59e199047aa8bf2cdd4401ae9f5f42b929e6/data/hyps/hyp.scratch-low.yaml#L6-L34

### Augmentation Previews

You can view the effect of your augmentation policy in your train_batch*.jpg images once training starts. These images will be in your train logging directory, typically `yolov5/runs/train/exp`:

`train_batch0.jpg` shows train batch 0 mosaics and labels:  
<img src="https://user-images.githubusercontent.com/26833433/131255960-b536647f-7c61-4f60-bbc5-cb2544d71b2a.jpg" width="800">


### YOLOv5 Albumentations Integration

[YOLOv5](https://github.com/ultralytics/yolov5) 🚀 is now fully integrated with [Albumentations](https://github.com/albumentations-team/albumentations), a popular open-source image augmentation package. Now you can train the world's best Vision AI models even better with custom Albumentations 😃! 

PR https://github.com/ultralytics/yolov5/pull/3882 implements this integration, which will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.3` is installed in your environment. See https://github.com/ultralytics/yolov5/pull/3882 for full details.

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

<img src="https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg" width="800">

Good luck 🍀 and let us know if you have any other questions!

**Comment by Carolinejone** [comment]
@glenn-jocher I can't thank you enough. You're my lifesaver. Your explanation is very effective and useful. I changed the model hyperparameters from low to high. Is it called fine-tuning? ( I admit that I had little knowledge about fine-tuning)
!python train.py --hyp hyp.scratch-high.yaml --img 640 --batch 64 --epochs 300 --data {dataset.location}/data.yaml --weights yolov5s6.pt --cache
This is the latest result I got for my Dataset-1. Does it overfit? Should I train more epochs?
I'll update the result for Dataset-2 when I finished running.
![confusion_matrix](https://user-images.githubusercontent.com/88965698/173544906-3f9fdd07-662e-41d6-a2b5-7b13283b2993.png)
![F1_curve](https://user-images.githubusercontent.com/88965698/173544932-4a570888-59fc-48ce-9eef-7f0900861646.png)
![labels](https://user-images.githubusercontent.com/88965698/173544950-ba8632a9-fa4e-4f0d-a91d-f1279c8f8d5e.jpg)
![P_curve](https://user-images.githubusercontent.com/88965698/173544969-b2d1430d-84b1-4491-965d-bc80687fa876.png)
![PR_curve](https://user-images.githubusercontent.com/88965698/173544988-db65b638-b7a3-408b-85db-886e5421d9de.png)
![R_curve](https://user-images.githubusercontent.com/88965698/173545002-0a69e0ad-ee4e-4a3d-93af-e1b00ea89d17.png)
![results](https://user-images.githubusercontent.com/88965698/173545018-8a5e0894-44cc-446f-948b-d619532aae37.png)

**Comment by glenn-jocher** [comment]
Seems like it’s now overfitting less, and you can now train longer

**Comment by Carolinejone** [comment]
@glenn-jocher Thank you so much. Now I've trained the Dataset-1 for 600 epochs and it starts overfitting after 500 epochs. (the model also stopped showing improvements after 484 epochs which I got the best results). So should I stop at epoch 500?
This is the updated result for Dataset-1. Your kind help is deeply appreciated.
![Capture](https://user-images.githubusercontent.com/88965698/173585530-e1f6a7ba-ec2b-469b-9bdf-552cd181ad43.JPG)
![confusion_matrix](https://user-images.githubusercontent.com/88965698/173585536-1b7b32b9-4ae0-4a41-926d-91594d78b671.png)
![F1_curve](https://user-images.githubusercontent.com/88965698/173585551-2f6b81d4-4f50-49e5-9059-2d3760d69bdd.png)
![labels](https://user-images.githubusercontent.com/88965698/173585562-9ea4fe9a-9c12-4a1d-9c59-cbf8fd1a3d6e.jpg)
![P_curve](https://user-images.githubusercontent.com/88965698/173585579-08e6a266-7ed6-481d-b71a-2db69ea0dad3.png)
![PR_curve](https://user-images.githubusercontent.com/88965698/173585592-3155e370-1e82-4821-a3ab-e2b680e2b58c.png)
![R_curve](https://user-images.githubusercontent.com/88965698/173585612-5fd23e9f-eb65-480e-b752-964f535abedb.png)
![results](https://user-images.githubusercontent.com/88965698/173585644-72eeca35-eb2e-4c1c-853d-fda46ff33a05.png)

**Comment by glenn-jocher** [comment]
@Carolinejone results look good!

**Comment by Carolinejone** [comment]
@glenn-jocher Thank you so much. I'll update the other results after training all datasets with different epochs. Again, your kind guidance is deeply appreciated.

**Comment by glenn-jocher** [comment]
@Carolinejone you're welcome! Feel free to reach out if you have any more questions. Good luck with your training, and I'm looking forward to seeing your results for the other datasets. Keep up the great work!

**Comment by pratikshac15** [comment]
@glenn-jocher 
Hello Glenn, I am not able to understand the background class and how it is calculated. Could you please explain me? also while calculating FN, FP, TN from the confusion matrix do I need to consider background class? 
Your reply would be appreciated. thank you in advance.

**Comment by glenn-jocher** [comment]
@pratikshac15 in YOLOv5, the background class refers to the class that represents the absence of any object in the image. When calculating FN (False Negative), FP (False Positive), and TN (True Negative) from the confusion matrix, these metrics are typically considered with respect to the classes that are being detected, and the background class is not factored into these calculations. Thank you for your question!

**Comment by CLaNLoRrD** [comment]
@glenn-jocher Hi Glenn, is there a brief summary of what each of the results graphs represents in the case of YOLOv5? I would like to explain the graphs in my thesis. Thank you in advance.
